checker/src/call_graph.rs

```diff
@@ -720,13 +684,175 @@ impl CallGraph {
+                        (SimpleType::Collection(t1_base), SimpleType::Base(t2_base)) => {
+                            if self.simple_type_eq(t1_base.as_ref(), t2_base.as_ref(), eq_map) {
+                                relations
+                                    .insert(TypeRelation::new_member(t1.to_owned(), t2.to_owned()));
+                            }
+                        }
+                        // Case 2: t < [t]
+                        (SimpleType::Base(t1_base), SimpleType::Collection(t2_base)) => {
+                            if self.simple_type_eq(t1_base.as_ref(), t2_base.as_ref(), eq_map) {
+                                relations
+                                    .insert(TypeRelation::new_member(t2.to_owned(), t1.to_owned()));
+                            }
+                        }
+                        // Case 3: [t] = [t] | t = t
+                        (SimpleType::Collection(t1_base), SimpleType::Collection(t2_base))
+                        | (SimpleType::Base(t1_base), SimpleType::Base(t2_base)) => {
+                            // Avoid trivial or redundant relations
+                            // - Trivial: EqType(a, a)
+                            // - Redundant: EqType(a, b) and EqType(b, a)
+                            if self.simple_type_eq(t1_base.as_ref(), t2_base.as_ref(), eq_map)
+                                && pairs.get(&(t1.as_ref(), t2.as_ref())).is_none()
+                                && pairs.get(&(t2.as_ref(), t1.as_ref())).is_none()
+                            {
+                                // Deterministic lexicographic ordering
+                                if t1 < t2 {
+                                    relations
+                                        .insert(TypeRelation::new_eq(t1.to_owned(), t2.to_owned()));
+                                } else {
+                                    relations
+                                        .insert(TypeRelation::new_eq(t2.to_owned(), t1.to_owned()));
+                                }
+                                pairs.insert((t1.as_ref(), t2.as_ref()));
+                                pairs.insert((t2.as_ref(), t1.as_ref()));
+                            }
+                        }
+                    }
+                }
+            }
+        }
+        relations
+    }
+
+    fn get_input_equivalences(
+        &self,
+        input_relations: &HashSet<TypeRelation>,
+    ) -> HashMap<Box<str>, Box<str>> {
+        let mut eq_map = HashMap::<Box<str>, Box<str>>::new();
+        for relation in input_relations.iter() {
+            match relation.kind {
+                TypeRelationKind::Eq => {
+                    eq_map.insert(relation.type1.to_owned(), relation.type2.to_owned());
+                    eq_map.insert(relation.type2.to_owned(), relation.type1.to_owned());
+                }
+                TypeRelationKind::Member => {}
+            }
+        }
+        eq_map
```

Comment:
This looks like filter_map to me.

Reply:
It doesn't quite work as a filter_map because we are inserting two values into the eq_map per one value in the type_relations. Thus, we'd need to run the filter_map twice, which seems less efficient to me.

Reply:
Try flat_map

Reply:
That worked, thanks!